api/hasura/actions/createUsers.ts

```diff
@@ -0,0 +1,155 @@
+import type { VercelRequest, VercelResponse } from '@vercel/node';
+
+import { authCircleAdminMiddleware } from '../../../api-lib/circleAdmin';
+import { ValueTypes } from '../../../api-lib/gql/__generated__/zeus';
+import { adminClient } from '../../../api-lib/gql/adminClient';
+import { errorResponseWithStatusCode } from '../../../api-lib/HttpError';
+import {
+  composeHasuraActionRequestBody,
+  createUsersBulkSchemaInput,
+} from '../../../src/lib/zod';
+
+async function handler(req: VercelRequest, res: VercelResponse) {
+  const {
+    input: { payload: input },
+  } = composeHasuraActionRequestBody(createUsersBulkSchemaInput).parse(
+    req.body
+  );
+
+  const { circle_id, users } = input;
+
+  const uniqueAddresses = [...new Set(users.map(u => u.address))];
+
+  // Check if bulk contains duplicates.
+  if (uniqueAddresses.length < users.length) {
+    return errorResponseWithStatusCode(
+      res,
+      {
+        message: `Users list contains duplicate address.`,
+      },
+      422
+    );
+  }
+
+  const { users: existingUsers } = await adminClient.query({
+    users: [
+      {
+        where: {
+          circle_id: { _eq: circle_id },
+          address: { _in: uniqueAddresses },
+        },
+      },
+      {
+        id: true,
+        address: true,
+        deleted_at: true,
+        starting_tokens: true,
+      },
+    ],
+  });
+
+  const usersToUpdate = existingUsers.map(eu => {
+    const updatedUser = users.find(u => u.address === eu.address);
+    return {
+      ...eu,
+      ...updatedUser,
+      give_token_remaining: updatedUser?.starting_tokens,
+    };
+  });
+
+  // Check if a user exists without a soft deletion conflicts with the new bulk.
+  const conflict = usersToUpdate.find(u => !u.deleted_at);
```

Comment:
we should compile the full list of addresses with conflicts before returning a response

Reply:
This is reasonable, but do you mean returning the whole list of `conflict` with the error message? if yes, Is there a way to include it which is easier to read in front-end.

Reply:
its good enough if you just join the array with a comma seperator within the `message`